### DISCUSSION TITLE
SPA mode and having access to nuxt instance and req/res

### DISCUSSION THREAD

**Discussion by JulienMaurice** [nuxt/nuxt#17371] [Questions]
Hello,

I'm making a plugin giving the possibilities to add httpHeader trought `nuxt.config.ts` and `composable`.

In SSR mode, I've been able to add to `ssrContext`.
But in SPA mode, `ssrContext` isn't usefull.

Before we had `addServerMiddleware`, with a callback function allowing us to access it. It's now depreciated

I tried with `addServerHandler`, to add a middleware with `(req, res)` or `(event)`, but I can't access nuxt instance in it, so i can't acces the option I gave to the plugin.

**Comment by Flanker72** [comment]
As described in the documentation, Nuxt plugin have the only argument passed - nuxtApp: https://v3.nuxtjs.org/guide/directory-structure/plugins#creating-plugins
Maybe, this will help you to reach the instance.

Also if you are going to implement the middleware, you may try to use `useNuxtApp()` composable. However, I don't know exactly if this is available in the middleware level.

**Reply by JulienMaurice** [reply]
Having the instance isn't a problem. Having access to the response or at least the httpheader of the response when SSR mode is off is.
In the definition of the module, or the defintion of the plugin, how can I have access to `req/res`.

I would like to add the httpHeader that give module options.

As you said, at the middleware level, we don't have access to `useNuxtApp()`
